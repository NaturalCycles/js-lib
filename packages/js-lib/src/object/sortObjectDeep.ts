/**
 * based on: https://github.com/IndigoUnited/js-deep-sort-object
 */
export function _sortObjectDeep<T>(o: T): T {
  // array
  if (Array.isArray(o)) {
    return o.map(_sortObjectDeep) as any
  }

  if (o && typeof o === 'object') {
    const r = {} as T
    for (const k of Object.keys(o).sort((a, b) => a.localeCompare(b)) as (keyof T)[]) {
      r[k] = _sortObjectDeep(o[k])
    }
    return r
  }

  return o
}
