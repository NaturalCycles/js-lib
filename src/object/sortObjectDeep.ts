import { _isObject } from '..'

/**
 * based on: https://github.com/IndigoUnited/js-deep-sort-object
 */
export function _sortObjectDeep<T>(o: T): T {
  // array
  if (Array.isArray(o)) {
    // eslint-disable-next-line unicorn/no-array-callback-reference
    return o.map(_sortObjectDeep) as any
  }

  if (_isObject(o)) {
    const out = {} as T

    Object.keys(o)
      .sort((a, b) => a.localeCompare(b))
      .forEach(k => {
        out[k as keyof T] = _sortObjectDeep(o[k as keyof T])
      })

    return out
  }

  return o
}
