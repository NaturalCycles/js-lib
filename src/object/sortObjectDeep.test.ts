import { expect, test } from 'vitest'
import { _sortObjectDeep } from './sortObjectDeep.js'

test('sortObjectDeep', () => {
  const o = {
    b: 2,
    a: 1,
    c: {
      d: {
        m: 4,
        j: 5,
        aa: 1,
      },
      m: 'sdf',
      dd: [3, 2, 1, 0, 55],
      ddd: ['b', 'c', 'a'],
    },
    aa: 6,
  }

  // console.log(sortObjectDeep(o))
  expect(_sortObjectDeep(o)).toMatchSnapshot()
})
