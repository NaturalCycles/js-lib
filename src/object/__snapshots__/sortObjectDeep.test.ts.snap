// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sortObjectDeep 1`] = `
{
  "a": 1,
  "aa": 6,
  "b": 2,
  "c": {
    "d": {
      "aa": 1,
      "j": 5,
      "m": 4,
    },
    "dd": [
      3,
      2,
      1,
      0,
      55,
    ],
    "ddd": [
      "b",
      "c",
      "a",
    ],
    "m": "sdf",
  },
}
`;
